syft/source/unindexed_directory_resolver.go

```diff
@@ -0,0 +1,565 @@
+
+func (u UnindexedDirectoryResolver) FileMetadataByLocation(_ Location) (FileMetadata, error) {
+	panic("FileMetadataByLocation unsupported")
+}
+
+func (u UnindexedDirectoryResolver) Write(filePath string, reader io.Reader) error {
+	if path.IsAbs(filePath) {
+		filePath = filePath[1:]
+	}
+	absPath := u.absPath(filePath)
+	return afero.WriteReader(u.fs, absPath, reader)
+}
+
+var _ FileResolver = (*UnindexedDirectoryResolver)(nil)
+var _ WritableFileResolver = (*UnindexedDirectoryResolver)(nil)
+
+func (u UnindexedDirectoryResolver) newLocation(filePath string, resolveLinks bool) *Location {
+	filePath = path.Clean(filePath)
+
+	virtualPath := ""
+	realPath := filePath
+
+	if resolveLinks {
+		paths := u.resolveLinks(filePath)
+		if len(paths) > 1 {
+			realPath = paths[len(paths)-1]
+			if realPath != path.Clean(filePath) {
+				virtualPath = paths[0]
+			}
+		}
+		if len(paths) == 0 {
+			// this file does not exist, don't return a location
+			return nil
+		}
+	}
+
+	return &Location{
+		Coordinates: Coordinates{
+			RealPath: realPath,
+		},
+		VirtualPath: virtualPath,
+	}
+}
+
+//nolint:gocognit
+func (u UnindexedDirectoryResolver) resolveLinks(filePath string) []string {
+	var visited []string
+
+	out := []string{}
+
+	resolvedPath := ""
+
+	parts := strings.Split(filePath, "/")
+	for i := 0; i < len(parts); i++ {
+		part := parts[i]
+		if resolvedPath == "" {
+			resolvedPath = part
+		} else {
+			resolvedPath = path.Clean(path.Join(resolvedPath, part))
+		}
+		resolvedPath = u.scrubResolutionPath(resolvedPath)
+		if resolvedPath == ".." {
+			resolvedPath = ""
+			continue
+		}
+
+		absPath := u.absPath(resolvedPath)
+		if slices.Contains(visited, absPath) {
+			return nil // circular links can't resolve
+		}
+		visited = append(visited, absPath)
+
+		fi, wasLstat, err := u.ls.LstatIfPossible(absPath)
+		if fi == nil || err != nil {
+			// this file does not exist
+			return nil
+		}
+
+		for wasLstat && u.isSymlink(fi) {
+			next, err := u.lr.ReadlinkIfPossible(absPath)
+			if err == nil {
+				if !path.IsAbs(next) {
+					next = path.Clean(path.Join(path.Dir(resolvedPath), next))
+				}
+				next = u.scrubResolutionPath(next)
+				absPath = u.absPath(next)
+				if slices.Contains(visited, absPath) {
+					return nil // circular links can't resolve
+				}
+				visited = append(visited, absPath)
+
+				fi, wasLstat, err = u.ls.LstatIfPossible(absPath)
+				if fi == nil || err != nil {
+					// this file does not exist
+					return nil
+				}
+				if i < len(parts) {
+					out = append(out, path.Join(resolvedPath, path.Join(parts[i+1:]...)))
+				}
+				if u.base != "" && path.IsAbs(next) {
+					next = next[1:]
+				}
+				resolvedPath = next
+			}
+		}
+	}
+
+	out = append(out, resolvedPath)
+
+	return out
+}
+
+func (u UnindexedDirectoryResolver) isSymlink(fi os.FileInfo) bool {
+	return fi.Mode().Type()&fs.ModeSymlink == fs.ModeSymlink
+}
+
+// ------------------------- fs.FS ------------------------------
+
+type fsFS struct {
```

Comment:
the name of `fsFS` isn't really clear here as to what it is and what it's for

Reply:
I would comment, but I expect this is more for @kzantow as I have no idea what that is 😁